File: py/tests/btest.py
import sys
import re
import numpy as np
import os.path

import poly_ops
import polydraw

DIFF_FAIL_SQUARE_SIZE = 5
DUMP_FAILURE_DIFF = False

# Technically, comments can appear anywhere, including inside 'tokens', not just
# around tokens, but we only need to open our own files, so this is good enough.
# The 'P4' at the start is omitted because it is checked separately.
RE_PBM_HEADER = re.compile(b"""
    (?:#.*[\n\r])*\\s(?:#.*[\n\r]|\\s)*
    (\\d+)
    (?:#.*[\n\r])*\\s(?:#.*[\n\r]|\\s)*
    (\\d+)
    (?:#.*[\n\r])*\\s""",re.VERBOSE)


class ImgFormatError(Exception):
    pass

def read_pbm(filename):
    with open(filename,'rb') as f_in:
        data = f_in.read()
    if not data.startswith(b'P4'):
        raise ImgFormatError('not a binary PBM file',filename)
    
    m = RE_PBM_HEADER.match(data,2)
    if not m:
        raise ImgFormatError('invalid PBM file',filename)
    width = int(m[1],base=10)
    height = int(m[2],base=10)
    return np.unpackbits(
        np.frombuffer(data,np.ubyte,offset=m.end()).reshape((height,(width+7)//8)),
        axis=1,
        count=width)

def write_pbm(filename,data):
    pbm_buffer = np.packbits(data,axis=1)
    with open(filename,'xb') as out:
        out.write(f'P4 {data.shape[1]} {data.shape[0]} '.encode('ascii'))
        out.write(pbm_buffer)

def has_square(data):
    if data.shape[0] < DIFF_FAIL_SQUARE_SIZE or data.shape[1] < DIFF_FAIL_SQUARE_SIZE:
        raise Exception('image too small for meaningful comparison')
    
    inner_w = data.shape[1]-DIFF_FAIL_SQUARE_SIZE+1
    inner_h = data.shape[0]-DIFF_FAIL_SQUARE_SIZE+1
    summed = np.ones((inner_h,inner_w),np.ubyte)

    for y in range(DIFF_FAIL_SQUARE_SIZE):
        for x in range(DIFF_FAIL_SQUARE_SIZE):
            summed &= data[y:y+inner_h,x:x+inner_w]
    
    return np.any(summed)

class TestCase:
    def __init__(self):
        self.set_data = ([],[])
        self.op_files = [None] * len(poly_ops.BoolOp)

class ParseError(Exception):
    pass

def require_whitespace_or_comment(line,lnum):
    line = line.lstrip(' \t\n\r')
    if line and not line.startswith("#"):
        raise ParseError(lnum)

def parse_tests(ipath):
    tests = [TestCase()]
    touched = False
    last_set = None
    partial_line = None

    re_array_assign_start = re.compile(r'[ \t]*:[ \t]*\[')
    re_array_assign_end = re.compile(r'([0-9 +\-]*)(\])[ \t\n\r]*(?:$|#)')
    re_op_assign = re.compile(r'[ \t]*:[ \t]*([0-9a-zA-Z]+)[ \t\n\r]*(?:$|#)')

    def end_array_assign(m,lnum):
        nonlocal touched, partial_line
        if not m:
            raise ParseError(lnum)
        partial_line.append(m[1])
        if m[2] is None: return
        tests[-1].set_data[last_set].append(np.fromstring(' '.join(partial_line),np.int_,sep=' ').reshape((-1,2)))
        touched = True
        partial_line = None

    with open(ipath) as ifile:
        for i,line in enumerate(ifile):
            if partial_line is not None:
                end_array_assign(re_array_assign_end.match(line),i)
            else:
                for s in poly_ops.BoolSet:
                    if line.startswith(s.name):
                        last_set = s
                        m = re_array_assign_start.match(line,len(s.name))
                        if not m:
                            raise ParseError(i)
                        partial_line = []
                        end_array_assign(re_array_assign_end.match(line,m.end()),i)
                        break
                else:
                    for op in poly_ops.BoolOp:
                        if line.startswith(op.name):
                            m = re_op_assign.match(line,len(op.name))
                            if not m:
                                raise ParseError(i)
                            tests[-1].op_files[op] = m[1]
                            touched = True
                            break
                    else:
                        if line.startswith("-"):
                            require_whitespace_or_comment(line.lstrip("-"),i)
                            if touched:
                                tests.append(TestCase())
                                touched = False
                        else:
                            require_whitespace_or_comment(line,i)

    if partial_line is not None:
        raise ParseError(-1)

    if not touched: tests.pop()
    return tests

def test_op(clip,rast,test,op,file_id,folder):
    target_img = read_pbm(os.path.join(folder,file_id + ".pbm"))
    test_img = np.zeros_like(target_img)

    clip.add_loops_subject(test.set_data[poly_ops.BoolSet.subject])
    clip.add_loops_clip(test.set_data[poly_ops.BoolSet.clip])

    rast.reset()
    rast.add_loops(clip.execute_flat(op))
    for x1,x2,y,wind in rast.scan_lines(test_img.shape[1],test_img.shape[0]):
        if wind > 0:
            test_img[y,x1:x2+1] = 1

    diff = test_img != target_img
    if has_square(diff):
        print(f'failure with operation "{op.name}" compared to file "{file_id}"',file=sys.stderr)
        if DUMP_FAILURE_DIFF:
            write_pbm(file_id + "_mine.pbm",test_img)
            write_pbm(file_id + "_diff.pbm",diff)
        return False
    return True

def run(datafile):
    tests = 0
    successes = 0

    cases = parse_tests(datafile)
    
    test_source = os.path.dirname(datafile)

    # make sure bitmap::has_square works
    assert not has_square(read_pbm(os.path.join(test_source,"discont.pbm")))
    assert has_square(read_pbm(os.path.join(test_source,"cont.pbm")))

    clip = poly_ops.Clipper()
    rast = polydraw.Rasterizer()

    for case in cases:
        has_one = False
        for op in poly_ops.BoolOp:
            if case.op_files[op] is not None:
                has_one = True
                successes += test_op(clip,rast,case,op,case.op_files[op],test_source)
                tests += 1

        if not has_one:
            print("warning: test entry found with nothing to test against",file=sys.stderr)
            continue;

    print(f"passed tests: {successes} out of {tests}",file=sys.stderr)

    return tests != successes

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("exactly one argument is required",file=sys.stderr)
        exit(2)
    exit(run(sys.argv[1]))

File: tests/btest.cpp
/* Test the algorithm by painting the output and comparing to pre-rendered
images of the same operations done per-pixel.

The results will not be exactly the same because the coordinates of the added
points are rounded to integers, thus a comparison is only considered a failure
when a continuous 5x5 pixel area doesn't match.

This program runs tests specified in an ASCII text file with the following
grammar (using the same notation as W3C:
https://www.w3.org/TR/REC-xml/#sec-notation):

Start       ::= (Line? (LineEnd Line?)*)?

Line        ::= Command S? Comment? | S Comment? | Comment

Command     ::= ArrayAssign | OpAssign | Divider

ArrayAssign ::= Set S? ":" S? "[" (S? Point (S Point))? S? "]"

OpAssign    ::= Operation S? ":" S? ID

Divider     ::= "-"+

Point       ::= Integer S Integer

ID          ::= [0-9a-zA-Z]+

LineEnd     ::= "\r"? "\n" | "\r"

Comment     ::= "#" [^\n\r]*

Set         ::= "subject" | "clip"

Operation   ::= "union" | "difference" | "intersection" | "xor" | "normalize"

Integer     ::= [+-]? [0-9]+

S           ::= [ \t]+

*/

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <algorithm>
#include <optional>
#include <string>
#include <filesystem>
#include <system_error>
#include <functional>
#include <iostream>
#include <charconv>

#include "bitmap.hpp"
#include "../include/poly_ops/polydraw.hpp"
#include "../include/poly_ops/base.hpp"
#include "../include/poly_ops/clip.hpp"
#include "../include/poly_ops/offset.hpp"


template<typename T> constexpr T MAX_IMG_DIM = 1000;
constexpr std::size_t MAX_INT_STR_SIZE = 9;

/* big enough to store a UUID if needed */
constexpr std::size_t MAX_ID_STR_SIZE = 36;

constexpr unsigned int DIFF_FAIL_SQUARE_SIZE = 5;

constexpr int OFFSET_ARC_STEP_SIZE = 3;

constexpr int TEST_ALL = -1;
constexpr int TEST_OFFSET = -2;


using coord_t = long;

#include "stream_output.hpp"

bool dump_failure_diff = false;

class parse_input_error : public std::exception {
public:
    const char *msg;
    int line;
    
    parse_input_error(const char *msg) : msg{msg}, line{-1} {}
    parse_input_error(const char *msg,unsigned int line) : msg{msg}, line{static_cast<int>(line)} {}

    const char *what() const noexcept override { return "bad input"; }

    void print(std::FILE *file) const noexcept {
        if(line >= 0) std::fprintf(file,msg,line);
        else std::fputs(msg,file);
    }
};

struct op_file_t {
    int offset;
    std::string file;
};

template<typename Coord> struct test_case {
    std::vector<std::vector<poly_ops::point_t<Coord>>> subject;
    std::vector<std::vector<poly_ops::point_t<Coord>>> clip;
    std::vector<op_file_t> op_files[static_cast<int>(poly_ops::bool_op::normalize)+1];
};

template<typename Coord> struct parse_state {
    std::vector<test_case<Coord>> tests;

    std::FILE *file;
    unsigned int line;
    int c;
    bool touched;

    int next_c() {
        c = std::getc(file);
        if(c == EOF && std::ferror(file)) throw std::system_error(errno,std::generic_category());
        return c;
    }

    void unexpected_char() {
        switch(c) {
        case EOF:
            throw parse_input_error{"unexpected end of file"};
        case '\n':
        case '\r':
            throw parse_input_error{"unexpected line end at line %i",line};
        default:
            throw parse_input_error{"unexpected character at line %i",line};
        }
    }

    void read_newline() {
        switch(c) {
        case '\n':
            line += 1;
            next_c();
            break;
        case '\r':
            line += 1;
            next_c();
            if(c == '\n') next_c();
            break;
        case EOF:
            break;
        default:
            unexpected_char();
        }
    }

    void skip_line() {
        for(;;) {
            switch(c) {
            case '\n':
            case '\r':
            case EOF:
                return;
            default:
                next_c();
                break;
            }
        }
    }

    void skip_whitespace(bool allow_newline=false) {
        for(;;) {
            switch(c) {
            case ' ':
            case '\t':
                next_c();
                break;
            case '#':
                skip_line();
                if(!allow_newline) return;
                break;
            case '\n':
                if(!allow_newline) return;
                line += 1;
                next_c();
                break;
            case '\r':
                if(!allow_newline) return;
                line += 1;
                next_c();
                if(c == '\n') next_c();
                break;
            default:
                return;
            }
        }
    }

    unsigned int read_uint() {
        char digits[MAX_INT_STR_SIZE];
        std::size_t str_size = 0;

        while(c >= '0' && c <= '9') {
            if(str_size == MAX_INT_STR_SIZE) throw parse_input_error{"too many digits in number at line %i",line};
            digits[str_size++] = char(c);
            next_c();
        }

        if(str_size == 0) unexpected_char();

        unsigned int value = 0;
        unsigned int factor = 1;
        while(str_size > 0) {
            --str_size;
            value += static_cast<unsigned int>(digits[str_size] - '0') * factor;
            factor *= 10;
        }
        return value;
    }

    int read_int() {
        bool neg = false;
        if(c == '-') {
            next_c();
            neg = true;
        } else if(c == '+') {
            next_c();
        }

        int value = static_cast<int>(read_uint());
        return neg ? -value : value;
    }

    std::string read_id() {
        char value[MAX_ID_STR_SIZE];
        std::size_t str_size = 0;

        while(c >= '0' && c <= '9') {
            if(str_size == MAX_ID_STR_SIZE) throw parse_input_error{"too many characters in ID at line %i",line};
            value[str_size++] = char(c);
            next_c();
        }
        return {value,str_size};
    }

    std::vector<poly_ops::point_t<Coord>> read_point_array() {
        std::vector<poly_ops::point_t<Coord>> points;

        if(c != '[') unexpected_char();
        next_c();

        poly_ops::point_t<Coord> p;
        for(;;) {
            skip_whitespace(true);
            if(c == ']') {
                next_c();
                return points;
            }
            p.x() = read_int();

            skip_whitespace(true);
            if(c == ']') throw parse_input_error{"array has odd number of items at line %i",line};
            p.y() = read_int();

            points.push_back(p);
        }
    }

    void read_field_name(const char *end) {
        while(*end) {
            if(*end++ != next_c()) {
                throw parse_input_error{"unknown field name on line %i",line};
            }
        }
        next_c();
        if((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')) {
            throw parse_input_error{"unknown field name on line %i",line};
        }
    }

    void read_colon() {
        skip_whitespace();
        if(c != ':') unexpected_char();
        next_c();
        skip_whitespace();
    }

    void read_op_entry(const char *end,poly_ops::bool_op op) {
        read_field_name(end);

        int offset = 0;
        if(c == '+' || c == '-') offset = read_int();

        read_colon();

        auto &file_set = tests.back().op_files[static_cast<int>(op)];
        auto i = std::ranges::lower_bound(file_set,offset,{},[](const op_file_t &o){ return o.offset; });
        if(i != file_set.end() && i->offset == offset) *i = {offset,read_id()};
        else file_set.emplace(i,offset,read_id());
        skip_whitespace();
        read_newline();
        touched = true;
    }

    void run(std::FILE *_file) {
        file = _file;
        line = 1;
        touched = false;
        next_c();
        tests.emplace_back();

        for(;;) {
            switch(c) {
            case 's':
                read_field_name("ubject");
                read_colon();
                tests.back().subject.push_back(read_point_array());
                touched = true;
                break;
            case 'c':
                read_field_name("lip");
                read_colon();
                tests.back().clip.push_back(read_point_array());
                touched = true;
                break;
            case 'u':
                read_op_entry("nion",poly_ops::bool_op::union_);
                break;
            case 'd':
                read_op_entry("ifference",poly_ops::bool_op::difference);
                break;
            case 'i':
                read_op_entry("ntersection",poly_ops::bool_op::intersection);
                break;
            case 'x':
                read_op_entry("or",poly_ops::bool_op::xor_);
                break;
            case 'n':
                read_op_entry("ormalize",poly_ops::bool_op::normalize);
                break;
            case '-':
                do {
                    next_c();
                } while(c == '-');
                skip_whitespace();
                read_newline();
                if(touched) {
                    tests.emplace_back();
                    touched = false;
                }
                break;
            case EOF:
                if(!touched) tests.pop_back();
                return;
            case ' ':
            case '\t':
                skip_whitespace();
                read_newline();
                break;
            case '\n':
            case '\r':
                read_newline();
                break;
            case '#':
                skip_line();
                read_newline();
                break;
            default:
                unexpected_char();
            }
        }
    }
};

const char *bool_op_names[] = {
    "union",
    "intersection",
    "xor",
    "difference",
    "normalize"};
constexpr int OP_COUNT = static_cast<int>(std::extent_v<decltype(bool_op_names)>);

bool test_op(
    poly_ops::clipper<coord_t> &clip,
    poly_ops::draw::rasterizer<coord_t> &rast,
    const test_case<coord_t> &test,
    poly_ops::bool_op op,
    const op_file_t &entry,
    const std::filesystem::path &folder)
{
    auto target_img = bitmap::from_pbm_file((folder / (entry.file + ".pbm")).c_str());
    bitmap test_img{target_img.width(),target_img.height()};
    test_img.clear();

    if(entry.offset) {
        add_offset_loops_subject(clip,test.subject,entry.offset,OFFSET_ARC_STEP_SIZE);
        add_offset_loops_clip(clip,test.clip,entry.offset,OFFSET_ARC_STEP_SIZE);
    } else {
        clip.add_loops_subject(test.subject);
        clip.add_loops_clip(test.clip);
    }

    auto clipped = clip.execute<false>(op);
    //for(auto loop : clipped) {
    //    std::cout << '[' << delimited(loop) << "]\n";
    //}

    rast.clear();
    rast.add_loops(clipped);
    rast.draw(
        test_img.width(),
        test_img.height(),
        std::bind_front(&bitmap::set_row,std::ref(test_img)),
        poly_ops::draw::fill_rule_t::positive);

    auto diff = test_img ^ target_img;
    if(diff.has_square(DIFF_FAIL_SQUARE_SIZE)) {
        std::fprintf(
            stderr,
            "failure with operation \"%s\" compared to file \"%s\"\n",
            bool_op_names[static_cast<int>(op)],
            entry.file.c_str());
        if(dump_failure_diff) {
            test_img.to_pbm_file((entry.file + "_mine.pbm").c_str());
            diff.to_pbm_file((entry.file + "_diff.pbm").c_str());
        }
        return false;
    }
    return true;
}

/* make sure bitmap::has_square works */
void check_square_finder(const std::filesystem::path &folder) {
    if(bitmap::from_pbm_file((folder / ("discont.pbm")).c_str()).has_square(DIFF_FAIL_SQUARE_SIZE)) {
        throw std::logic_error{"bitmap::has_square is faulty; found square where none should exist"};
    }
    if(!bitmap::from_pbm_file((folder / ("cont.pbm")).c_str()).has_square(DIFF_FAIL_SQUARE_SIZE)) {
        throw std::logic_error{"bitmap::has_square is faulty; did not find square where one should exist"};
    }
}

bool str_to_bool(const char *str) {
    if(!str) return false;
    int x;
    const char *end = str+std::strlen(str);
    auto r = std::from_chars(str,end,x);
    return r.ptr == end && x;
}

int main(int argc,char *argv[]) {
    dump_failure_diff = str_to_bool(std::getenv("DUMP_FAILURE_DIFF"));

    if(argc != 2 && argc != 3) {
        std::fprintf(stderr,"exactly one or two arguments are required\n");
        return 2;
    }

    int operation = TEST_ALL;
    if(argc > 2) {
        if(std::strcmp("offset",argv[2]) == 0) {
            operation = TEST_OFFSET;
            goto found;
        }
        for(int i=0; i<OP_COUNT; ++i) {
            if(std::strcmp(bool_op_names[i],argv[2]) == 0) {
                operation = i;
                goto found;
            }
        }
        if(std::strcmp("all",argv[2]) != 0) {
            std::fprintf(stderr,"invalid operation name\n");
            return 2;
        }
        found: ;
    }

    std::size_t tests = 0;
    std::size_t successes = 0;

    try {
        parse_state<coord_t> state;

        std::FILE *file = std::fopen(argv[1],"rb");
        if(!file) throw std::system_error(errno,std::generic_category());

        {
            file_closer closer{file};
            state.run(file);
        }

        auto test_source = std::filesystem::path(argv[1]).remove_filename();

        check_square_finder(test_source);

        poly_ops::clipper<coord_t> clip;
        poly_ops::draw::rasterizer<coord_t> rast;

        for(auto &test : state.tests) {
            bool has_one = false;
            for(int i=0; i<OP_COUNT; ++i) {
                if(!test.op_files[i].empty()) {
                    has_one = true;
                    if(operation > -1 && operation != i) continue;
                    for(auto &op_file : test.op_files[i]) {
                        if(operation == TEST_ALL || (op_file.offset != 0) == (operation == TEST_OFFSET)) {
                            successes += test_op(clip,rast,test,static_cast<poly_ops::bool_op>(i),op_file,test_source);
                            tests += 1;
                        }
                    }
                }
            }
            if(!has_one) {
                std::fputs("warning: test entry found with nothing to test against\n",stderr);
            }
        }
        std::fprintf(stderr,"passed tests: %zu out of %zu\n",successes,tests);
    } catch(const parse_input_error &e) {
        e.print(stderr);
        std::fputc('\n',stderr);
        return 2;
    } catch(const std::exception &e) {
        std::fputs(e.what(),stderr);
        std::fputc('\n',stderr);
        return 3;
    }

    return successes != tests;
}
